# Cas d'usage n°1

Après récupération d'un jeu de données quelconque (ici les industries classées pour la protection de l'environnement = ICPE de la région Hauts-de-France), l'objectif est de retrouver les codes communes manquants.

## Import des modules

In [1]:
import os
import pandas as pd
from requests_cache import CachedSession
from tqdm import tqdm

from french_cities import find_city

## Constitution du jeu de données : récupération des ICPE de la région Hauts-de-France

In [2]:
code_region = "32"
page_size = 1000

s = CachedSession()
r = s.get(
    "https://georisques.gouv.fr/api/v1/installations_classees",
    params={"page": "1", "page_size": page_size, "region": code_region},
)
number_pages = r.json()["total_pages"]
for x in tqdm(range(number_pages), desc="querying georisques"):
    try:
        data
    except NameError:
        data = []
    else:
        r = s.get(
            "https://georisques.gouv.fr/api/v1/installations_classees",
            params={
                "page": x + 1,
                "page_size": page_size,
                "region": code_region,
            },
        )
    finally:
        data += r.json()["data"]
    if not r.json()["next"]:
        break
data = pd.DataFrame(data)

querying georisques:  92%|████████████████████████████████████████████████████████▎    | 12/13 [00:00<00:00, 18.74it/s]


## Analyse du jeu de données obtenu

In [3]:
data

raisonSociale  \
0                                   SCEA FERME LABALETTE   
1                           ISDI de Saint-Laurent-Blangy   
2                                 RECYCL'ELECTRONIC SARL   
3                                ISDI -TCPA (PATINIER A)   
4                                               LAV'ALIM   
...                                                  ...   
12175  PARC EOLIEN DE NOYERS ST MARTIN (repowering ex...   
12176                                   LEFEBVRE Mathieu   
12177                                            CAR PRO   
12178                                LEROY MERLIN FRANCE   
12179                                     RICHARD Arnaud   

                           adresse1 codePostal codeInsee  \
0                Lieu-dit Le Dièvre      62860     62739   
1          Zone des Trois Fontaines      62223     62753   
2      2 Bis Rue <charles Duquesnoy      62270     62361   
3                    rue de la Gare      62470     62197   
4       Zone Industrielle Le Royeux      02430     02340   
...                             ...        ...       ...   
12175       Lieu-dit Le Cornouiller      60480     60634   
12176                 rue de Ferfay      62190     62028   
12177         12 GRANDE RUE DE VAUX      60390     60063   
12178             RTE DE ST QUENTIN      80330     80489   
12179           136 rue Jean Jaurès      59224     59589   

                    commune  longitude   latitude  bovins  porcs  volailles  \
0        Sains-lès-Marquion   3.083774  50.176359   False  False      False   
1      Saint-Laurent-Blangy   2.809392  50.303387   False  False      False   
2                   Frévent   2.293696  50.267633   False  False      False   
3        Camblain-Châtelain   2.429339  50.478130   False  False      False   
4                    Gauchy   3.293073  49.822421   False  False      False   
...                     ...        ...        ...     ...    ...        ...   
12175   Noyers-Saint-Martin   2.287214  49.548878   False  False      False   
12176                  Ames   2.419674  50.540442   False  False      False   
12177      BERNEUIL-EN-BRAY   2.041738  49.355651   False  False      False   
12178              LONGUEAU   2.367384  49.867366   False  False      False   
12179                Thiant   3.451233  50.311974   False  False      False   

       ...  serviceAIOT          regime  \
0      ...    DREAL HdF  Enregistrement   
1      ...    DREAL HdF        Non ICPE   
2      ...    DREAL HdF  Autres régimes   
3      ...    DREAL HdF        Non ICPE   
4      ...    DREAL HdF  Autres régimes   
...    ...          ...             ...   
12175  ...    DREAL HdF    Autorisation   
12176  ...    DREAL HdF        Non ICPE   
12177  ...    DREAL HdF        Non ICPE   
12178  ...    DREAL HdF        Non ICPE   
12179  ...    DREAL HdF        Non ICPE   

                                               rubriques  \
0      [{'numeroRubrique': '2760', 'nature': 'Install...   
1                                                     []   
2                                                     []   
3                                                     []   
4                                                     []   
...                                                  ...   
12175  [{'numeroRubrique': '2980', 'nature': 'Eolienn...   
12176                                                 []   
12177                                                 []   
12178                                                 []   
12179                                                 []   

                              inspections  \
0                                      []   
1      [{'dateInspection': '2019-02-04'}]   
2      [{'dateInspection': '2015-09-29'}]   
3      [{'dateInspection': '2016-12-06'}]   
4      [{'dateInspection': '2020-02-24'}]   
...                                   ...   
12175                                  []   
12176  [{'dateInspection': '2024-07-02'}]   
12177  [{

## ICPE dépourvues de codes commune INSEE

In [4]:
data.codeInsee.isnull().value_counts().to_frame("Nombre de codes manquants :")

Nombre de codes manquants :
codeInsee                             
False                            11926
True                               254

##### A date du 21/07/2024, le jeu de données est constitué de 12180 lignes et 31 colonnes, dont 254 codes communes INSEE manquants. Pourtant, d'autres champs
sont disponibles et exploitables pour retrouver les communes manquantes :
* des champs adresse ;
* un libellé de commune ;
* des coordonnées géographiques et référentiel de projection ;
* un code postal.

## Complétion des données avec `french-cities`
### Configuration de l'API INSEE

In [5]:
os.environ["insee_key"] = "********************"
os.environ["insee_secret"] = "********************"

### Extraction des données manquantes

In [6]:
missing = data[data.codeInsee.isnull()].copy()

Au besoin, vérifier que le système de projection des coordonnées est en EPSG 2154 (lorsqu'il est connu) :

In [7]:
missing[["systemeCoordonneesAIOT"]].fillna("est manquant").value_counts().to_frame()

count
systemeCoordonneesAIOT       
est manquant              239
2154                       15

### Concaténation des adresses :

In [8]:
cols = [f"adresse{x}" for x in range(1, 4)]
addresses = (
    missing[cols[0]]
    .str.cat(missing[cols[1:]], sep=" ", na_rep="")
    .str.replace(" +", " ", regex=True)
    .str.strip(" ")
)
missing["adresse"] = addresses

### Recherche des communes manquantes à l'aide de `french-cities` :

In [9]:
filled = find_city(
    missing,
    year="last",
    x="coordonneeXAIOT",
    y="coordonneeYAIOT",
    epsg=2154,
    city="commune",
    dep=False,
    address="adresse",
    postcode="codePostal",
    use_nominatim_backend=False,
    field_output="newCodeInsee",
)
filled

raisonSociale                    adresse1  \
17                 SARL FINANCIERE VARET  Lieu-dit "les Huit Cannes"   
501                          GSMC Market     ZAC DU PONT SANS PAREIL   
1429                    THESEE FORMATION        Domaine des Vivirets   
3632                           ALTEA MSO              ZAC du Plateau   
3708          LES FRUITS ROUGES AISNE SA            1 rue Jean Bodin   
...                                  ...                         ...   
12062                  CRAPET Jean-Louis          le fond du Marquoi   
12079  PRESSING DES PORTES DE L'AVESNOIS         RTE DE VALENCIENNES   
12080                          VAL PRESS             72 AV DE DENAIN   
12122                            NORBLOC         50 route de Tournai   
12127                        AUTO SELECT       394 B RTE DE FEIGNIES   

      codePostal codeInsee              commune  longitude   latitude  bovins  \
17         62118       NaN              Fampoux  -2.279526  45.502618   False   
501        62610       NaN       BOIS EN ARDRES        NaN        NaN   False   
1429       60490       NaN          MARQUEGLISE        NaN        NaN   False   
3632       02200       NaN               Ploisy        NaN        NaN   False   
3708       02000       NaN                 Laon  -1.363081  -5.983856   False   
...          ...       ...                  ...        ...        ...     ...   
12062      62550       NaN  Bailleul-lès-Pernes  -5.525510  42.559454   False   
12079      59530       NaN           LE QUESNOY  -0.001143   0.001046   False   
12080      59300       NaN         VALENCIENNES  -0.001105   0.001474   False   
12122      59226       NaN             Lecelles  -0.001220  -0.001001   False   
12127      59600       NaN             MAUBEUGE  -0.002286   0.000888   False   

       porcs  volailles  ...  \
17     False      False  ...   
501    False      False  ...   
1429   False      False  ...   
3632   False      False  ...   
3708   False      False  ...   
...      ...        ...  ...   
12062  False      False  ...   
12079  False      False  ...   
12080  False      False  ...   
12122  False      False  ...   
12127  False      False  ...   

                                               rubriques  \
17     [{'numeroRubrique': '2760', 'nature': 'Install...   
501                                                   []   
1429                                                  []   
3632                                                  []   
3708   [{'numeroRubrique': '1511', 'nature': 'Entrepô...   
...                                                  ...   
12062                                                 []   
12079                                                 []   
12080                                                 []   
12122                                                 []   
12127                                                 []   

                                             inspections  \
17                                                    []   
501                   [{'dateInspection': '2017-11-10'}]   
1429   [{'dateInspection': '2022-03-22', 'fichierInsp...   
3632                                                  []   
3708   [{'dateInspection': '2023-06-12', 'fichierInsp...   
...                                                  ...   
12062  [{'dateInspection': '2023-10-06', 'fichierInsp...   
12079  [{'dateInspection': '2023-10-11', 'fichierInsp...   
12080  [{'dateInspection': '2023-10-11', 'fichierInsp...   
12122  [{'dateInspection': '2024-02-16', 'fichierInsp...   
12127  [{'dateInspection': '2024-03-06', 'fichierInsp...   

                                 documentsHorsInspection             date_maj  \
17     [{'identifiantFichier': 'xwOI57mkibOLwaTaqUaWl...  2024-06-19/19-12-56   
501                                                   []  2024-06-12/12-42-23   
1429                                                  []                  NaN   
3632   [{'identifiantFichier': '10a4a1

### Réinjection les codes manquants dans le dataframe complet :

In [10]:
data = data.join(filled[["newCodeInsee"]])
data["codeInsee"] = data["codeInsee"].combine_first(data["newCodeInsee"])
data = data.drop("newCodeInsee", axis=1)

### Evaluation des résultats

In [11]:
data.codeInsee.isnull().value_counts().to_frame("Nombre de codes manquants :")

Nombre de codes manquants :
codeInsee                             
False                            12179
True                                 1

### Données toujours manquantes

In [12]:
data[data.codeInsee.isnull()]

raisonSociale           adresse1 codePostal  \
8441  ATELIERS DES EPICES ET CONDIMENTS  19 RUE DE LA GARE      62360   

     codeInsee          commune  longitude  latitude  bovins  porcs  \
8441       NaN  PONT DE BRIQUES        NaN       NaN   False  False   

      volailles  ...  serviceAIOT    regime  rubriques  \
8441      False  ...    DREAL HdF  Non ICPE         []   

                             inspections documentsHorsInspection  \
8441  [{'dateInspection': '2015-01-12'}]                      []   

                 date_maj               adresse2 codeNaf           siret  \
8441  2024-05-30/10-47-45  SAINT ETIENNE AU MONT     NaN  53197729600023   

      adresse3  
8441       NaN  

[1 rows x 31 columns]

A date du 21/07/2024, une seule commune n'a pas été trouvée.

Effectivement, dans ce cas de figure, le lieu-dit (PONT DE BRIQUES) et la commune (SAINT ETIENNE AU MONT) ont été inversés : ceci explique que le score de la base adresse nationale n'ait pas été jugé suffisamment bon pour que le résultat de Saint-Etienne-au-Mont puisse être retenu...

Si cette fois, on décide d'utiliser l'API Nominatim en dernier recours, le code devient :

In [13]:
# On isole la(es) ligne(s) manquante(s)
missing = data[data.codeInsee.isnull()].copy()

# On concatène de nouveau les champs adresses :
cols = [f"adresse{x}" for x in range(1, 4)]
addresses = (
    missing[cols[0]]
    .str.cat(missing[cols[1:]], sep=" ", na_rep="")
    .str.replace(" +", " ", regex=True)
    .str.strip(" ")
)
missing.loc[:, "adresse"] = addresses

# Et on spécifie l'usage de Nominatim
missing = find_city(
    missing,
    year="last",
    x=False,
    y=False,
    dep=False,
    epsg=2154,
    city="commune",
    address="adresse",
    postcode="codePostal",
    use_nominatim_backend=True,
    field_output="newCodeInsee",
)

Utilisons pynsee pour récupérer la liste des communes pour contrôler le résultat (et pas simplement son code commune) :

In [14]:
# Présentation des résultats
from pynsee.localdata import get_area_list
cities = get_area_list("communes", date="*")
missing["newCodeInsee"].to_frame().merge(cities, left_on="newCodeInsee", right_on="CODE")

newCodeInsee   CODE                                                URI  \
0        62746  62746  http://id.insee.fr/geo/commune/adc106e4-5055-4...   

  AREA_TYPE DATE_CREATION            TITLE_SHORT DETERMINER_TYPE  \
0   Commune    1943-01-01  Saint-Étienne-au-Mont               0   

                   TITLE DATE_DELETION  
0  Saint-Étienne-au-Mont           NaN

💡 Nota : l'exécution de Nominatim ne conduit pas systématiquement au même résultat (et même parfois ne produit pas de résultat). Cela n'est pas totalement absurde, le hameau manquant étant à cheval sur plusieurs communes. Les résultats fournis restent  généralement pertinents.